### PR TITLE
replace hierarchy searches for implementations of closed interfaces with literal implementation matches

### PR DESCRIPTION
**Pull request by richardstartin**
When a framework's interface is effectively closed and not open for external implementation, we don't need to support _any_ implementation, just the names of the implementations. There are more places we can do this safely (e.g. rabbitmq `Channel`) but I believe all of the changes here are interfaces not implemented externally.

`Tracing` instrumentations now have optionally two type matching concepts:

* `shortCutMatcher()` - disabled by default, fast, will match all known implementations, may miss custom implementations. Arguably provides the best experience for a balance of coverage against startup performance, but if we made it default it would be annoying for users where things "just work" in ways we don't know about and they need to set a property to maintain tracing and metrics. Enabled by `-Ddd.integration.<instrumentation name>.matching.shortcut.enabled=true` e.g. `-Ddd.integration.httpclient.matching.shortcut.enabled=true` 
* `hierarchyMatcher()` - enabled by default, will search the type hierarchy for all implementations, but it's slow. 

Instrumenters don't need to make this distinction: they can just override `Tracing.typeMatcher`. Some interfaces are not suitable for short cut matching - e.g. runnables, listeners, promises, anything with a template method, anything where the API involves inheriting or implementing a type, so we should reserve the short cuts for closed interfaces.

I enabled short cuts by default for
* Dropwizard Views
* Hibernate (has the added benefit of distinguishing between versions prior to muzzle, reducing the number of failing muzzle match attempts)

I'm open to suggestions about whether these are wise choices.

### CODE REVIEW

File: dd-java-agent/instrumentation/apache-httpclient-4/src/main/java/datadog/trace/instrumentation/apachehttpclient/ApacheHttpClientInstrumentation.java [hunk 1]
```diff
@@ -41,7 +41,19 @@ public ApacheHttpClientInstrumentation() {
...
+        "org.apache.http.impl.client.AutoRetryHttpClient",
+        "org.apache.http.impl.client.CloseableHttpClient",
+        "org.apache.http.impl.client.ContentEncodingHttpClient",
+        "org.apache.http.impl.client.DecompressingHttpClient",
+        "org.apache.http.impl.client.DefaultHttpClient",
+        "org.apache.http.impl.client.InternalHttpClient",
+        "org.apache.http.impl.client.MinimalHttpClient",
+        "org.apache.http.impl.client.SystemDefaultHttpClient",
+        "com.netflix.http4.NFHttpClient",
+        "com.amazonaws.http.apache.client.impl.SdkHttpClient");
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/apache-httpclient-4/src/main/java/datadog/trace/instrumentation/apachehttpclient/ApacheHttpClientInstrumentation.java:56]: This one is obviously not a closed interface but I think we can track all common implementations down. I find it unlikely that this interface would be implemented in application code.

File: dd-java-agent/instrumentation/twilio/src/test/groovy/test/TwilioClientTest.groovy [hunk 2]
```diff
@@ -27,6 +27,12 @@ import static datadog.trace.agent.test.utils.TraceUtils.runUnderTrace
 
 class TwilioClientTest extends AgentTestRunner {
 
+  @Override
+  protected void configurePreAgent() {
+    System.setProperty("dd.integration.httpclient.hierarchy.matching.enabled", "true")
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/twilio/src/test/groovy/test/TwilioClientTest.groovy:32]: Why is this necessary?

Comment (richardstartin) [reply dd-java-agent/instrumentation/twilio/src/test/groovy/test/TwilioClientTest.groovy:32]: Because the test mocks HttpClient, but we don't need to match mocks by default, only a finite set of commonly used implementations from the library itself and a couple of other libraries. If a user wants to pay in startup time to match literally any HttpClient implementation, including mocks, they can do that, by setting this flag - see the PR description.

Comment (tylerbenson) [reply dd-java-agent/instrumentation/twilio/src/test/groovy/test/TwilioClientTest.groovy:32]: yeah, I understand that part.  I didn't understand why it was needed in this specific test.

**Review by bantonsson — APPROVED** [review]
Good idea. Haven't verified all the implementations, but it looks reasonable.

**Comment by richardstartin** [comment]
This change is saving ~400ms to load 10k classes in a dropwizard application

File: dd-java-agent/instrumentation/hibernate/core-4.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_3/ProcedureCallInstrumentation.java [hunk 3]
```diff
@@ -47,8 +46,13 @@ public ProcedureCallInstrumentation() {
...
   @Override
-  public ElementMatcher<TypeDescription> typeMatcher() {
-    return implementsInterface(named("org.hibernate.procedure.ProcedureCall"));
+  public ElementMatcher<TypeDescription> shortCutMatcher() {
+    return named("org.hibernate.procedure.internal.ProcedureCallImpl");
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return named("org.hibernate.procedure.ProcedureCall");
```
Comment (mcculls) [inline dd-java-agent/instrumentation/hibernate/core-4.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_3/ProcedureCallInstrumentation.java:55]: missing `implementsInterface`?

File: dd-java-agent/instrumentation/hibernate/core-4.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_3/SessionInstrumentation.java [hunk 4]
```diff
@@ -58,8 +59,19 @@ public SessionInstrumentation() {
   }
 
   @Override
-  public ElementMatcher<TypeDescription> typeMatcher() {
-    return implementsInterface(named("org.hibernate.SharedSessionContract"));
+  public ElementMatcher<? super TypeDescription> hierarchyMatcher() {
+    return implementsInterface(named("org.hibernate.Session"));
```
Comment (mcculls) [inline dd-java-agent/instrumentation/hibernate/core-4.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_3/SessionInstrumentation.java:63]: wouldn't this change mean `org.hibernate.internal.AbstractSharedSessionContract` is now missed in the `hierarchyMatcher` mode, since it only implements `SharedSessionContract` and not `Session`?

(this class is listed in the `shortCutMatcher` so I assume it's important)

Comment (richardstartin) [reply dd-java-agent/instrumentation/hibernate/core-4.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_3/SessionInstrumentation.java:63]: Sorry this was missed in some of the churn earlier, will fix it. Thanks for spotting.

**Comment by richardstartin** [comment]
I will look at finding a simple way to run these instrumentations in both matching modes to catch mistakes

File: dd-java-agent/instrumentation/hibernate/core-4.0/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_0/SessionFactoryInstrumentation.java [hunk 5]
```diff
@@ -33,8 +33,13 @@
...
   @Override
-  public ElementMatcher<TypeDescription> typeMatcher() {
-    return implementsInterface(named("org.hibernate.SessionFactory"));
+  public ElementMatcher<TypeDescription> shortCutMatcher() {
+    return named("org.hibernate.internal.SessionFactoryImpl");
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return implementsInterface(named("org.hibernate.impl.SessionFactory"));
```
Comment (mcculls) [inline dd-java-agent/instrumentation/hibernate/core-4.0/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_0/SessionFactoryInstrumentation.java:42]: is this name right? originally this was `org.hibernate.SessionFactory`

browsing through the hibernate code I don't see any `SessionFactory` in the `org.hibernate.impl` package

Comment (richardstartin) [reply dd-java-agent/instrumentation/hibernate/core-4.0/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_0/SessionFactoryInstrumentation.java:42]: Yep, I think we need to test with each mode enabled to catch copy and paste errors like this. The interface is in `org.hibernate`.

File: dd-java-agent/instrumentation/hibernate/core-4.0/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_0/SessionInstrumentation.java [hunk 6]
```diff
@@ -46,8 +46,14 @@
...
-  public ElementMatcher<TypeDescription> typeMatcher() {
-    return implementsInterface(named("org.hibernate.SharedSessionContract"));
+  public ElementMatcher<TypeDescription> shortCutMatcher() {
+    return namedOneOf(
+        "org.hibernate.internal.SessionImpl", "org.hibernate.internal.StatelessSessionImpl");
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return implementsInterface(named("org.hibernate.Session"));
```
Comment (mcculls) [inline dd-java-agent/instrumentation/hibernate/core-4.0/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_0/SessionInstrumentation.java:56]: Will this cover `org.hibernate.internal.StatelessSessionImpl`?  I don't think that type actually implements `org.hibernate.Session` which is why  the original matcher used `org.hibernate.SharedSessionContract`

File: dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/SessionFactoryInstrumentation.java [hunk 7]
```diff
@@ -43,8 +43,13 @@
...
   @Override
-  public ElementMatcher<TypeDescription> typeMatcher() {
-    return implementsInterface(named("org.hibernate.SessionFactory"));
+  public ElementMatcher<TypeDescription> shortCutMatcher() {
+    return named("org.hibernate.impl.SessionFactoryImpl");
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return implementsInterface(named("org.hibernate.impl.SessionFactory"));
```
Comment (mcculls) [inline dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/SessionFactoryInstrumentation.java:52]: As below, shouldn't this be `org.hibernate.SessionFactory` ?

File: dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/SessionInstrumentation.java [hunk 8]
```diff
@@ -50,9 +50,13 @@
...
-  public ElementMatcher<TypeDescription> typeMatcher() {
-    return implementsInterface(
-        namedOneOf("org.hibernate.Session", "org.hibernate.StatelessSession"));
+  public ElementMatcher<TypeDescription> shortCutMatcher() {
+    return namedOneOf("org.hibernate.impl.SessionImpl", "org.hibernate.impl.StatelessSessionImpl");
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return implementsInterface(named("org.hibernate.Session"));
```
Comment (mcculls) [inline dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/SessionInstrumentation.java:59]: Will this cover `org.hibernate.impl.StatelessSessionImpl`? I don't think that type actually implements `org.hibernate.Session` which is why the original matcher included `org.hibernate.StatelessSession `

File: dd-java-agent/instrumentation/hibernate/core-4.0/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_0/SessionInstrumentation.java (hunk 6 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/hibernate/core-4.0/src/main/java/datadog/trace/instrumentation/hibernate/core/v4_0/SessionInstrumentation.java:56]: Yes, sorry, this got mangled when introducing the fallback mechanism. I now have tests for both modes which catches manual errors like this.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java [hunk 9]
```diff
@@ -313,14 +313,42 @@ public boolean isApplicable(Set<TargetSystem> enabledSystems) {
...
+    @Override
+    public ElementMatcher<? super TypeDescription> typeMatcher() {
+      if (shortCutEnabled) {
+        return shortCutMatcher();
+      }
+      return hierarchyMatcher();
+    }
+
+    public ElementMatcher<? super TypeDescription> shortCutMatcher() {
+      throw new IllegalStateException("quickMatcher not implemented");
```
Comment (mcculls) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Instrumenter.java:346]: s/quickMatcher/shortCutMatcher/

**Review by mcculls — APPROVED** [review]

**Comment by richardstartin** [comment]
@mcculls I have added tests which would have caught the mistakes you found by inspection. I've squashed it all for the sake of git history though.

File: dd-java-agent/instrumentation/hibernate/core-4.0/src/test/groovy/HierarchyMatcherCriteriaTest.groovy [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class HierarchyMatcherCriteriaTest extends CriteriaTest {
+  @Override
+  protected void configurePreAgent() {
+    super.configurePreAgent()
+    injectSysConfig("dd.integration.hibernate.matching.shortcut.enabled", "false")
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/hibernate/core-4.0/src/test/groovy/HierarchyMatcherCriteriaTest.groovy:5]: 👍

**Review by bantonsson — APPROVED** [review]

**Review by charliegracie — APPROVED** [review]

**Review by tylerbenson — APPROVED** [review]
